runtime/jsTest/src/kotlinx/serialization/DynamicSerializerTest.kt

```diff
@@ -0,0 +1,338 @@
+
+    @Serializable
+    data class WithChar(val a: Char)
+
+    @Serializable
+    data class AllTypes(
+        val b: Byte,
+        val s: Short,
+        val i: Int,
+        val f: Float,
+        val d: Double,
+        val c: Char,
+        val B: Boolean,
+        val S: String
+    )
+
+    @Serializable
+    data class EnumWrapper(val e: Color)
+
+    @Serializable
+    sealed class Sealed {
+        @Serializable
+        data class One(val string: String) : Sealed()
+    }
+
+    @Serializable
+    class WithJsName(@JsName("b") val a: String)
+
+    @Serializable
+    class WithSerialName(@SerialName("b") val a: String)
+
+    @Serializable
+    enum class Color {
+        RED,
+        GREEN,
+        @SerialName("red")
+        WITH_SERIALNAME_red
+    }
+
+    @Serializable
+    class MyFancyClass(val value: String) {
+
+        companion object : KSerializer<MyFancyClass> {
+
+            override val descriptor: SerialDescriptor = StringDescriptor.withName("MyFancyClass")
+            override fun serialize(encoder: Encoder, obj: MyFancyClass) {
+                encoder.encodeString("fancy ${obj.value}")
+            }
+
+            override fun deserialize(decoder: Decoder): MyFancyClass {
+                return MyFancyClass(decoder.decodeString().removePrefix("fancy "))
+            }
+        }
+    }
+
+    private inline fun <reified T : Any> assertDynamicForm(
+        data: T,
+        serializer: SerializationStrategy<T>? = null,
+        noinline assertions: ((T, dynamic) -> Unit)? = null
+    ) {
+        val effectiveSerializer = serializer ?: EmptyModule.getContextualOrDefault(T::class)
+        val serialized = DynamicObjectSerializer().serialize(effectiveSerializer, data)
+        assertions?.invoke(data, serialized)
+        assertEquals(
+            Json.stringify(effectiveSerializer, data),
+            JSON.stringify(serialized),
+            "JSON.stringify representation must be the same"
+        )
+    }
+
+    @Test
+    fun dynamicSimpleTest() {
+        assertDynamicForm(Data(42)) { data, serialized ->
+            assertEquals(data.a, serialized.a)
+        }
+
+        assertDynamicForm(WithChar('c')) { data, serialized ->
+            assertEquals(data.a.toString(), serialized.a)
+        }
+
+        assertDynamicForm(AllTypes(1, 2, 3, 4.0f, 5.0, 'c', true, "string"))
+
+    }
+
+    @Test
+    fun wrappedObjectsTest() {
+        assertDynamicForm(DataWrapper("a string", Data(42))) { data, serialized ->
+            assertEquals(data.s, serialized.s)
+            assertNotNull(serialized.d)
+            assertEquals(data.d?.a, serialized.d.a)
+        }
+    }
+
+    @Test
+    fun nullsTest() {
+        val data = DataWrapper("a string", null)
+
+        val serialized = DynamicObjectSerializer(encodeDefaults = true).serialize(data)
+        assertNull(serialized.d)
+        assertFalse(js("""Object.keys(serialized).includes("d")"""), "should omit null properties")
```

Comment:
Idea of 'omitting defaults' is that we should omit _default property_ value, not _null_. Since the `DataWrapper` has declaration `val d: Data? = Data(1)`, we should omit d only if it is equal to `Data(1)`. That's automatically detected by generated serializer and then `shouldEncodeElementDefault` is called.